mobile/src/main/java/org/openhab/habdroid/model/OpenHABWidget.java

```diff
@@ -10,293 +10,329 @@
+    public abstract String iconPath();
+    public abstract Type type();
+    @Nullable
+    public abstract String url();
+    @Nullable
+    public abstract OpenHABItem item();
+    @Nullable
+    public abstract OpenHABLinkedPage linkedPage();
+    public abstract List<OpenHABWidgetMapping> mappings();
+    @Nullable
+    public abstract String encoding();
+    @Nullable
+    public abstract Integer iconColor();
+    @Nullable
+    public abstract Integer labelColor();
+    @Nullable
+    public abstract Integer valueColor();
+    public abstract int refresh();
+    public abstract float minValue();
+    public abstract float maxValue();
+    public abstract float step();
+    public abstract String period();
+    public abstract String service();
+    @Nullable
+    public abstract Boolean legend();
+    public abstract int height();
+
+    public boolean hasMappings() {
+        List<OpenHABWidgetMapping> mappings = mappings();
+        return mappings != null && !mappings.isEmpty();
     }
 
-    public boolean childrenHasNonlinkedPages() {
-        if (this.hasChildren()) {
-            for (OpenHABWidget w : this.getChildren()) {
-                if (!w.hasLinkedPage())
-                    return true;
+    abstract Builder toBuilder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+        public abstract Builder id(String id);
+        public abstract Builder parentId(@Nullable String parentId);
+        public abstract Builder label(@Nullable String label);
+        public abstract Builder icon(@Nullable String icon);
+        public abstract Builder iconPath(String iconPath);
+        public abstract Builder type(Type type);
+        public abstract Builder url(@Nullable String url);
+        public abstract Builder item(@Nullable OpenHABItem item);
+        public abstract Builder linkedPage(@Nullable OpenHABLinkedPage linkedPage);
+        public abstract Builder mappings(List<OpenHABWidgetMapping> mappings);
+        public abstract Builder encoding(@Nullable String encoding);
+        public abstract Builder iconColor(@Nullable Integer iconColor);
+        public abstract Builder labelColor(@Nullable Integer labelColor);
+        public abstract Builder valueColor(@Nullable Integer valueColor);
+        public abstract Builder refresh(int refresh);
+        public abstract Builder minValue(float minValue);
+        public abstract Builder maxValue(float maxValue);
+        public abstract Builder step(float step);
+        public abstract Builder period(String period);
+        public abstract Builder service(String service);
+        public abstract Builder legend(@Nullable Boolean legend);
+        public abstract Builder height(int height);
+
+        public OpenHABWidget build() {
+            // Consider a minimal refresh rate of 100 ms
+            refresh(Math.max(refresh(), 100));
+            // Default period to 'D'
+            if (period() == null || period().isEmpty()) {
```

Comment:
TextUtils.isEmpty() ?

Reply:
I think @FlorianSW previously asked (in a comment in #562 IIRC) for TextUtils not to be used in base code because it doesn't work for the tests.

Reply:
Right, it's an Android method. However we could write your own method in Utils.

Reply:
I'd leave this up for later and replace all such occurrences in our code. Alternatively, we could mock this method accordingly in our tests.